crates/builder/src/input/dotgraph.rs

```diff
@@ -265,23 +203,332 @@ where
+                    *count += 1;
+                }
+
+                frequency
+            })
+            .collect_into_vec(&mut frequencies);
+
+        Self {
+            frequencies,
+            phantom: PhantomData::default(),
+        }
+    }
+}
+
+pub struct NodeLabelIndex<Label, NI>(Csr<Label, NI, ()>)
+where
+    NI: Idx,
+    Label: Idx;
+
+impl<Label, NI> NodeLabelIndex<Label, NI>
+where
+    NI: Idx,
+    Label: Idx,
+{
+    pub fn from_stats<F>(node_count: NI, label_stats: LabelStats<NI, Label>, label_func: F) -> Self
+    where
+        Label: Hash,
+        F: Fn(NI) -> Label + Send + Sync,
+    {
+        (node_count, label_stats, label_func).into()
+    }
+
+    pub fn nodes(&self, label: Label) -> &[NI] {
+        self.0.targets(label)
+    }
+}
+
+impl<Label, NI, F> From<(NI, LabelStats<NI, Label>, F)> for NodeLabelIndex<Label, NI>
+where
+    NI: Idx,
+    Label: Idx + Hash,
+    F: Fn(NI) -> Label + Send + Sync,
+{
+    fn from((node_count, label_stats, label_func): (NI, LabelStats<NI, Label>, F)) -> Self {
+        let LabelStats {
+            label_count,
+            max_label,
+            label_frequency,
+            ..
+        } = label_stats;
+        // Prefix sum: We insert the offset entries one index to the right and
+        // increment the offset of the next label during insert. That way we'll
+        // end up with the correct offsets after inserting into `nodes` in the
+        // next loop.
+        let mut offsets = Vec::with_capacity(label_count.index() + 1);
+        offsets.push(Label::zero());
+
+        let mut total = Label::zero();
+        for label in Label::zero().range_inclusive(max_label) {
+            offsets.push(total);
+            total += Label::new(*label_frequency.get(&label).unwrap_or(&0));
+        }
+
+        let mut offsets = ManuallyDrop::new(offsets);
+        let (ptr, len, cap) = (offsets.as_mut_ptr(), offsets.len(), offsets.capacity());
+
+        // SAFETY: Label and Label::Atomic have the same memory layout
+        let offsets = unsafe {
+            let ptr = ptr as *mut Atomic<Label>;
+            Vec::from_raw_parts(ptr, len, cap)
+        };
+
+        let mut nodes = Vec::<Target<NI, ()>>::with_capacity(node_count.index());
+        let nodes_ptr = SharedMut::new(nodes.as_mut_ptr());
+
+        (0..node_count.index()).into_par_iter().for_each(|node| {
+            let label = label_func(NI::new(node));
+            let next_label = label + Label::new(1);
+            let offset = Label::get_and_increment(&offsets[next_label.index()], Acquire);
+            // SAFETY: There is exactly one thread that writes at `offset.index()`.
+            unsafe {
+                nodes_ptr
+                    .add(offset.index())
+                    .write(Target::new(NI::new(node), ()));
+            }
+        });
+
+        // SAFETY: The `labels` vec has `node_count` length and we performed an
```

Comment:
The test says `labels` vec, the code says `nodes` vec :)

Reply:
And the `doc` is wrong since it's actually `nodes` in the vec and not labels :)